crates/red_knot_python_semantic/src/types.rs

```diff
@@ -151,6 +165,73 @@ pub(crate) fn definitions_ty<'db>(
     }
 }
 
+/// Build a declared type from a [`DeclarationsIterator`].
+///
+/// Will return a union if there is more than one declaration, or at least one plus the possibility
+/// of undeclared.
+///
+/// If undeclared is a possibility, `Unknown` type will be part of the return type.
+///
+/// # Panics
+/// Will panic if there are no declarations and no possibility of undeclared. This is a logic
+/// error, as any symbol with zero live declarations clearly may (must!) be undeclared.
+fn declarations_ty<'db>(
+    db: &'db dyn Db,
+    declarations: DeclarationsIterator<'_, 'db>,
+) -> DeclaredType<'db> {
+    let may_be_undeclared = declarations.may_be_undeclared();
+    let decl_types = declarations.map(|declaration| declaration_ty(db, declaration));
+
+    let mut all_types = (if may_be_undeclared {
+        Some(Type::Unknown)
+    } else {
+        None
+    })
+    .into_iter()
+    .chain(decl_types);
+
+    let first = all_types.next().expect(
+        "declarations_ty must not be called with zero declarations and no may-be-undeclared.",
+    );
+
+    let mut conflicting: Vec<Type<'db>> = vec![];
+    let declared_ty = if let Some(second) = all_types.next() {
+        let mut builder = UnionBuilder::new(db).add(first);
+        for other in [second].into_iter().chain(all_types) {
+            if !first.is_equivalent_to(db, other) {
+                conflicting.push(other);
+            }
+            builder = builder.add(other);
+        }
+        builder.build()
+    } else {
+        first
+    };
+    DeclaredType {
+        declared_ty,
+        conflicting: if conflicting.is_empty() {
+            TypeList(conflicting.into())
+        } else {
+            TypeList([first].into_iter().chain(conflicting).collect())
+        },
+    }
+}
+
+#[derive(Debug)]
+struct DeclaredType<'db> {
+    declared_ty: Type<'db>,
+    conflicting: TypeList<'db>,
+}
+
+#[derive(Debug)]
+struct TypeList<'db>(Box<[Type<'db>]>);
+
+impl<'db> TypeList<'db> {
+    fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+}
```

Comment:
We can implement `Display` for `Box<[Type]>` if the `Display` type is the only reason for defining `TypeList`. 

Should `TupleType` and `FunctionType::decorators` use `TypeList`?

Reply:
We can't implement `Display` for `Box<[Type]>` directly, because we need a `db`. But I added a `DisplayTypeArray` intermediate struct and a `TypeArrayDisplay` trait with a `display(&self, db) -> DisplayTypeArray` method, and then we can implement that trait for `Box<[Type]>`. I think it works out pretty well, looking forward to your review :)